Static to static calls

Schaeff opened the pull request:
This PR further separates virtual machines from constrained machines:
- introduces a new machine statement `link` which enables calling into another machine
- reintroduces the machine statement `operation` which enables defining a low-level entry point
- reduce functions to operations in romgen
- reduce extrenal instructions to links in romgen

As a result of this PR, virtual machines are completely reduced to constrained machines (before, external instructions survived until airgen. They are now turned into links first)

Reading the book on this PR is recommended, `mdbook serve book` locally.

leonardoalt commented:
tests are failing on book examples

Schaeff commented on book/src/asm/registers.md:
````diff
@@ -43,4 +43,18 @@ However, if we want the instruction to accept any write register as input, we us
+{{#include ../../../test_data/asm/book/asm_function_after.asm.ignore}}
+```
+
+>>>>>>> e0effa1 (implement static-to-static calls)
````
oops

Schaeff commented:
> tests are failing on book examples

This is because witgen is not supported for links yet. I removed the link to make the example pass and added a TODO.

leonardoalt commented on test_data/asm/book/operations_and_links.asm:
```diff
@@ -0,0 +1,37 @@
+    operation main<0> x, y -> z;
+
+    // on every row, add in the Adder
+    link 1 x, y -> z = adder.add;
```
Was this 1 automatically generated? What's the asm that got compiled to this?

Schaeff replied:
No. This is a user-defined link, see book.

Schaeff replied:
I added more details in this source file.

leonardoalt commented on book/src/asm/functions.md:
```diff
@@ -1,18 +1,16 @@
 # Functions
 
-Machine functions are the entry points to a machine. They can be called from another machine or from the outside.
+Functions are the entry points to a virtual machine. They can be called from another machine or from the outside.
```
What does outside mean here actually, besides another machine?

Schaeff replied:
From a snark verifier basically, for the entry point.

leonardoalt commented on book/src/asm/instructions.md:
```diff
@@ -1,6 +1,6 @@
 # Instructions
 
-Instructions are declared as part of a powdr asm machine. Their inputs and outputs are [assignment registers](./registers.md) as well as labels. Once defined, they can be called by any function of this machine.
+Instructions are declared as part of a powdr virtual machine. Their inputs and outputs are [assignment registers](./registers.md) as well as labels. Once defined, they can be called by any function of this machine.
```
```suggestion
Instructions are declared as part of a powdr virtual machine. Their inputs and outputs are [assignment registers](./registers.md) as well as labels. Once defined, they can be called by any function in this machine.
```

leonardoalt commented on book/src/asm/machines.md:
```diff
@@ -2,45 +2,45 @@
 ## Submachines
 
-Machines can have submachines which they access by defining [external instructions](./instructions.md). They are declared as follows:
+Machines can have submachines which they access by defining [external instructions](./instructions.md) or [links](./links.md). They are declared as follows:
```
Wait, what's the difference between external instructions and links?

Schaeff replied:
Instructions are a virtual machine concept. Links are a constrained machine concept. Links rely on a user-defined boolean flag and user-defined columns. External instructions rely on an instruction name and assignment registers, and get compiled down to links where the link flag is the instruction flag (introduced internally). The inputs and outputs are the same as in the instruction declaration as assignment registers compile down to committed columns of the same name.

leonardoalt commented on book/src/architecture/README.md:
```diff
@@ -0,0 +1,15 @@
+# Architecture
+
+powdr applies a number of steps in order to reduce a powdr asm program into a PIL file.
```
```suggestion
powdr applies a number of steps in order to reduce a powdr-asm program into PIL.
```

leonardoalt commented on book/src/architecture/README.md:
````diff
@@ -0,0 +1,15 @@
+```ascii
+            ┌────────────┐           ┌──────────┐
+            │            │           │          │
+ powdr asm  │            │ AIR graph │          │  PIL
````
```suggestion
 powdr-asm  │            │ AIR graph │          │  PIL
```

leonardoalt commented on book/src/architecture/compiler.md:
```diff
@@ -0,0 +1,25 @@
+
+## Virtual machine reduction
+
+The first step is to reduce virtual machines to constrained machines. This step is run on all machines, it simply does not affect constrained machines.
```
```suggestion
The first step is to reduce virtual machines to constrained machines. This step is run on all machines and does not affect constrained machines.
```

leonardoalt commented on book/src/architecture/linker.md:
```diff
@@ -0,0 +1,21 @@
+- a list of PIL statements
+- a list of links to other machine instances
+
+A linker is used to turn this tree of AIRs into a structure we can generate proofs about. This currently takes the form of a single PIL file.
```
At this point it's not clear what the tree of AIRs is.

leonardoalt replied:
```suggestion
A linker is used to turn this tree of AIRs into a structure that is used to generate proofs. This currently takes the form of a single PIL file.
```

leonardoalt commented on book/src/architecture/linker.md:
```diff
@@ -0,0 +1,21 @@
+    3. Add the constraints to the namespace
+    4. Turn the links into lookups and add them to the namespace
+
+The result is a monolithic AIR where each machine instance is a namespace, where all namespaces have the same degree and links between instances are guaranteed by lookup identities.
```
and permutations?

Schaeff replied:
No, only lookups. The permutation argument for memory is internal to a virtual machine, i.e the memory machine is not a separate machine.

leonardoalt reviewed (commented):
Review for docs and tests

leonardoalt commented:
> This is because witgen is not supported for links yet. I removed the link to make the example pass and added a TODO.

ok, still failing tho

Schaeff commented on analysis/README.md:
```diff
@@ -510,21 +506,21 @@ pol constant p_reg_write_X_A = [0, 0, 0, 0, 0] + [0]*;
-// instr nothing with params  links to function "nothing" with id 4 with params  in object at location "main_sub" with latch "instr_return" and function_id "_function_id"
+// instr_identity  X -> Y links to function "identity" with id 2 with params  _input_0 -> _output_0 in object at location "main_sub" with latch "instr_return" and operation_id "_operation_id"
+// instr_one   -> Y links to function "one" with id 4 with params   -> _output_0 in object at location "main_sub" with latch "instr_return" and operation_id "_operation_id"
+// instr_nothing  links to function "nothing" with id 3 with params  in object at location "main_sub" with latch "instr_return" and operation_id "_operation_id"
```
There is no syntax for links in object files, so they are just referred to as comments. They are not expressible as PIL code unlike the rest of the object.

Schaeff replied:
Lookups are one way to encode the links. Links are turned into lookups in the linker. Here we are before the linker, where we have a bunch of independent tables, and some links between them, but no specific implementation of the links.

Schaeff commented:
> > This is because witgen is not supported for links yet. I removed the link to make the example pass and added a TODO.
> 
> ok, still failing tho

Didn't commit file 🤦

leonardoalt commented on book/src/architecture/compiler.md:
```diff
@@ -0,0 +1,29 @@
+
+Let's define AIR as a data structure with only these elements.
+
+Starting from the main machine type, we instanciate a tree of AIR by walking its tree of submachines, instantiating each machine as an AIR.
```
```suggestion
Starting from the main machine type, we instantiate a tree of AIR by walking its tree of submachines, instantiating each machine as an AIR.
```

Schaeff replied:
every time

leonardoalt commented on book/src/architecture/compiler.md:
```diff
@@ -0,0 +1,29 @@
+
+Let's define AIR as a data structure with only these elements.
+
+Starting from the main machine type, we instanciate a tree of AIR by walking its tree of submachines, instantiating each machine as an AIR.
```
or actually, I'd say this
```suggestion
Starting from the main machine's type, we create a tree of AIR objects by traversing its submachines, recursively instantiating each machine as an AIR.
```

Schaeff replied:
done

leonardoalt commented on ast/src/parsed/asm.rs:
```diff
@@ -44,42 +44,51 @@ impl Params {
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub struct CallableRef {
+    pub instance: String,
+    pub callable: String,
```
If this is a string, shouldn't the type checker check that the called operations has the correct arity?

Schaeff replied:
Yes it should. There is a lot more that can be checked in the type checker in general. I think it can be a different PR.

leonardoalt replied:
ok, can you add a TODO in the type checker and open an issue

Schaeff replied:
added TODO to the checked AST (this is the parsed one where this doesnt have to be enforced)

leonardoalt commented on ast/src/asm_analysis/display.rs:
```diff
@@ -47,23 +48,32 @@ where
 
+impl<T: Display> Display for LinkDefinitionStatement<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "link {} {} = {};", self.flag, self.params, self.to)
```
doesn't it have the arrow too?

Schaeff replied:
The arrow is included in the params : inputs -> outputs

georgwiese commented on pil_analyzer/src/pil_analyzer.rs:
```diff
@@ -348,7 +348,7 @@ impl<T: FieldElement> PILContext<T> {
             .definitions
             .insert(name.clone(), (poly, value))
             .is_none();
-        assert!(is_new, "{name} aready defined.");
```
🍻

Schaeff commented on ast/src/asm_analysis/mod.rs:
```diff
@@ -199,26 +213,322 @@ impl<T> FromIterator<Batch<T>> for FunctionStatements<T> {
+    pub name: &'a str,
+    /// a reference to the symbol
+    pub symbol: &'a CallableSymbol<T>,
+}
```
`CallableRef` is `some_machine.some_operation_or_function` as strings.
`CallableSymbol` is the definition of an operation or function.

Schaeff commented on ast/src/asm_analysis/mod.rs:
```diff
@@ -50,20 +53,31 @@ impl RegisterTy {
+    pub params: Params,
+    /// the callable to invoke when the flag is on
+    pub to: CallableRef,
+}
```
This is not calling anything, it's the declaration or either function or operation.

georgwiese commented on airgen/src/lib.rs:
```diff
@@ -118,23 +116,21 @@ impl<'a, T: FieldElement> ASMPILConverter<'a, T> {
-            .iter()
-            .all(|i| matches!(i.instruction.body, InstructionBody::External(..))));
+        // machines should only have constraints, operations and links at this point
+        assert!(input.instructions.is_empty());
```
```suggestion
        assert!(input.instructions.is_empty());
        assert!(input.function_definitions().next().is_none());
```

Schaeff replied:
This is already taken care of by `input.callable.is_only_operations()` down below

georgwiese reviewed (commented):
I mostly looked through the documentation, examples and the compiled PIL. Let's a realistic example and test it with #507! 💯

Schaeff commented on test_data/asm/book/operations_and_links.asm:
```diff
@@ -0,0 +1,38 @@
+    // link 1 x, y -> z = adder.add;
+
+    constraints {
+        col fixed operation_id = [0]*;
+        col fixed latch = [1]*;
```
Yes, it's the general case, as you must define how the constrained machine can be accessed from the outside. Removing would require some magic which takes the id of the only operation here (`main`) and sets the operation_id to that. For the latch, I can't see a way around it being user defined. In general I would stick with this explicit way for now.

Schaeff commented on test_data/asm/keccak.asm:
```diff
@@ -2,7 +2,7 @@
 
 // ================= binary/bitwise instructions =================
 
-machine Binary(latch, function_id) {
```
Hmm this file does not seem to be used anywhere

Schaeff replied:
added to release tests as it's used in benchmarks

Schaeff commented on book/src/asm/operations.md:
```diff
@@ -0,0 +1,13 @@
+# Operations
```
There is a case for links in VMs: the program lookup can be seen as an unconditional link. This would be used only internally now and I'm not sure how to expose it to the user yet. For operations I can't see a case right now and it seems unlikely there will be one.

Schaeff commented on test_data/asm/block_to_block.asm:
```diff
@@ -0,0 +1,42 @@
+
+    Arith arith;
+
+    // return `4*x + 4*y`, adding twice locally and twice externally
```
fixed

Schaeff commented on analysis/README.md:
```diff
@@ -4,15 +4,15 @@ This crate is where most of the compilation pipeline lives. It takes a parse tre
-| internal instructions   | yes     | no          |
+| operation_id            | no      | yes         |
+| functions               | yes     | no          |
+| instructions            | yes     | no          |
```
Good point but no, links are allowed in VMs

georgwiese replied:
```suggestion
| instructions            | yes     | no          |
| links                   | yes     | yes         |
```

This way then? ;) I think links should be mentioned here, otherwise it's the only major concept that's not.

Schaeff commented on book/src/linker/README.md:
```diff
@@ -1,21 +1 @@
 # Linker
```
removed already

Schaeff commented on book/src/asm/links.md:
```diff
@@ -0,0 +1,12 @@
+# Links
+
+Links enable a constrained machine to call into another machine.
```
I added more details in the code block

Schaeff commented on test_data/asm/block_to_block.asm:
```diff
@@ -0,0 +1,42 @@
+// calls two operations in a submachine whose interface is different: one is `x, y, z` while the other one is `z, x, y`
```
fixed

Schaeff commented on analysis/README.md:
````diff
@@ -438,17 +435,16 @@ machine Main(instr_return, _function_id) {
+
 }
 ```
 
````
fixed

Schaeff commented on analysis/README.md:
````diff
@@ -565,11 +561,11 @@ For our example, it introduces two namespaces (one per machine instance) with th
 ```
 namespace main(16);
-pol commit _function_id;
+pol commit _operation_id;
````
#514

leonardoalt commented on analysis/README.md:
```diff
@@ -101,9 +101,9 @@ Rom generation generates a single ROM for each virtual machine using the followi
 - Replace references to the function arguments by references to these input registers.
 - Pad all return statements with zeroes up to the number of output registers.
-- Inline all functions, adding a label before each one.
+- Inline all functions arbitrarily ordered by name, adding a label before each one.
```
```suggestion
- Inline all function bodies, arbitrarily sorted by name, adding a label before each one.
```
Otherwise it's not clear to me what "inline all functions" mean

leonardoalt commented on analysis/README.md:
```diff
@@ -321,13 +319,17 @@ Airgen takes machines which are only left with constraints and external instruct
+machine DifferentSignatures(instr_return, _operation_id) {
+        operation identity<2> _input_0 -> _output_0;
+        operation nothing<3>;
+        operation one<4>  -> _output_0;
```
```suggestion
        operation one<4> -> _output_0;
```
was this auto generated?

Schaeff replied:
yes

leonardoalt commented on ast/src/parsed/display.rs:
```diff
@@ -513,3 +539,61 @@ impl Display for UnaryOperator {
+                }],
+            }),
+        };
+        assert_eq!(in_out.to_string(), "abc: ty -> abc: ty");
```
maybe also a test with multiple inputs/outputs?

Schaeff replied:
done

leonardoalt reviewed (commented):
Looks good, small comments only.

Schaeff commented on test_data/asm/block_to_block.asm:
```diff
@@ -0,0 +1,47 @@
+    // return `3*x + 3*y`, adding twice locally and twice externally
+    operation main<0>;
+
+    link instr_add x, y -> z = arith.add;
```
agreed, syntactic change welcome

Schaeff commented on linker/src/lib.rs:
```diff
@@ -207,31 +205,31 @@ mod test {
 
     #[test]
-    pub fn compile_empty_vm() {
+    fn compile_empty_vm() {
```
sure

chriseth commented on compiler/tests/asm.rs:
```diff
@@ -81,7 +81,24 @@ fn vm_to_block_unique_interface() {
+#[test]
+#[ignore = "Too slow"]
+fn keccak() {
+    let f = "keccak.asm";
```
This file does not exist.
